daffodil-lib/src/main/scala/org/apache/daffodil/xml/DaffodilXMLLoader.scala

```diff
@@ -538,47 +401,226 @@ trait SchemaAwareLoaderMixin {
+    // The setEntityResolver(resolver) does not work.
+    //
+    xrdr.setProperty("http://apache.org/xml/properties/internal/entity-resolver", resolver)
+    p
+  }
+
+  /**
+   * This is the common routine called by all the load calls to actually
+   * carry out the loading of the schema.
+   *
+   * There are no calls to this in Daffodil code base as of this writing, but
+   * the base class scala.xml.factory.XMLLoader calls it.
+   *
    * Does (optional) validation,
+   *
+   * @param source The URI for the XML document which may be a XML or DFDL schema, or just XML data.
+   * @param optSchemaURI Optional URI for XML schema for the XML source document.
+   * @param addPositionAttributes True to add dafint:file dafint:line attributes to all elements.
+   *                              Defaults to false.
+   * @param normalizeCRLFtoLF True to normalize CRLF and isolated CR to LF. This should usually be true,
+   *                          but some special case situations may require preservation of CRLF/CR.
+   * @return an scala.xml.Node (Element actually) which is the document element of the source.
    */
-  def load(source: DaffodilSchemaSource): scala.xml.Node = {
-    if (doValidation) {
-      val inputSource = source.newInputSource()
-      val xercesNode = xercesAdapter.load(inputSource) // validates
-      inputSource.getByteStream().close()
-
-      if (xercesNode == null) return null
-      // Note: we don't call xercesAdapter.validateSchema(source)
-      // here, because this is an XML loader, not necessarily
-      // just a DFDL schema loader. So for example the doValidation flag
-      // above could be telling us to validate a TDML file or not.
+  def load(source: DaffodilSchemaSource,
+    optSchemaURI: Option[URI],
+    addPositionAttributes: Boolean = false,
+    normalizeCRLFtoLF: Boolean = true): scala.xml.Node = {
+    //
+    // First we invoke the validator to explicitly validate the XML against
+    // the XML Schema (not necessarily a DFDL schema), via the
+    // javax.xml.validation.Validator's validate method.
+    //
+    optSchemaURI.foreach { schemaURI =>
+
+      val validator = XercesValidator.fromURIs(Seq(schemaURI))
+      val inputStream = source.uriForLoading.toURL.openStream()
+      validator.validateXML(inputStream, errorHandler)
+      inputStream.close()
+      //
+      // Next we have to invoke a regular xerces loader, setup for validation
+      // because that will actually interpret things like xsi:schemaLocation attributes
+      // of the root element.
+      //
+      // scopeStack.push(TopScope) // not in scala xml v2.0.0
+      val parser = parserFromURI(optSchemaURI)
+      val xrdr = parser.getXMLReader()
+      val saxSource = scala.xml.Source.fromSysId(source.uriForLoading.toString)
+      // parser.parse(source.uriForLoading.toURL.openStream(), this)
+      xrdr.parse(saxSource)
+      // scopeStack.pop() // not in scala xml v2.0.0
```

Comment:
It would be good to document exactly what additional errors this second validation catches that the first validation does not. Without this numerous negative tests failed. 

At least do this: comment out this check, and list what the tests are that fail.

Reply:
Tried this. Tests that fail are:
TestTDMLRunner.test_tdmlNamespaces1 - does not get a TDMLException it is expecting which is specified in the TDML file using xsi:schemaLocation. 

This is not being checked at all without this added validation step. This is why the 2nd check is needed.
Comments in the code have been updated to indicate this. 

These tests are expecting a "DocType is disallowed" exception, and without that 2nd check, they get a NPE. 
I've fixed that so that the TDML runner doesn't throw a NPE any more. 

TestDisallowDocType.test_TDMLFileMustNotHaveDocType
TestDisallowDocType.test_dfdlSchemaMustNotHaveDocType
TestDisallowDocType.test_configMustNotHaveDocType